crates/bevy_ecs/src/observer/runner.rs

```diff
@@ -41,11 +51,17 @@ pub(super) fn observer_system_runner<E: Event, B: Bundle, S: ObserverSystem<E, B
     }
     state.last_trigger_id = last_trigger;
 
+    // SAFETY: Caller ensures `trigger_ptr` is castable to `&mut E::Trigger<'_>`
+    // This is enforced by the safety requirements of `Trigger::trigger`
+    // See the safety discussion on `Trigger::trigger` for more details.
+    let trigger: &mut E::Trigger<'_> = unsafe { trigger_ptr.deref_mut() };
+
     let on: On<E, B> = On::new(
-        // SAFETY: Caller ensures `ptr` is castable to `&mut T`
-        unsafe { ptr.deref_mut() },
-        propagate,
-        observer_trigger,
+        // SAFETY: Caller ensures `ptr` is castable to `&mut E`
+        unsafe { event_ptr.deref_mut() },
+        observer,
+        trigger,
+        trigger_context,
     );
```

Comment:
The tricky part here is that we're still constructing a `On<'a, 'a, E, B>`, most notably with the same lifetime `'a`, due to passing it to an `ObserverSystem` which takes it as a `SystemInput` (and that's forced to use one lifetime only).

We are hence casting `trigger_ptr` to a `&'a mut E::Trigger<'a>` (key detail: with the same lifetime twice), and the caller _cannot really guarantee this is sound_. It's really not for any type `E::Trigger<'a>` that actually uses its lifetime.

IMO it's important to document this here since it's a key and very subtle detail of this whole safety story.